### ISSUE TITLE
[Coroutines] Add a verifier for coroutines

### ISSUE THREAD

**Issue by ChuanqiXu9** [llvm/llvm-project#60863]
Currently the compiler will crash directly if the input is not valid. It is not friendly and confusing to people. We should better add a verifier for coroutines. Here is an example: https://github.com/llvm/llvm-project/issues/60713

**Comment by Endilll** [comment]
#30850 and friends showcase crashes while using coroutine builtins. While those builtins are not user-facing, I don't think crashing is a reasonable reaction to their misuse. 

If the verifier is not easy to implement (which I suspect is the case), then maybe we should ship a simple on-by-default stack-overflow-like diagnostic when coroutine builtins are used outside of known standard library headers and modules? Something along the lines of `Usage of coroutine builtins detected in user code. They are fragile and can cause compiler to crash.` At some point in the future we can replace it with proper verifier as a matter of quality of implementation.

CC @AaronBallman

**Comment by ChuanqiXu9** [comment]
I agree with you generally but I feel the priority may be low. The most related issue reports happen before 2020, which is in a stage we are still exploring and experimenting the methods to implement C++20 coroutines.

**Comment by AaronBallman** [comment]
FWIW, I agree with @Endilll that crashing is not a reasonable behavior for the builtins and I also agree with @ChuanqiXu9 that the priority is kind of low. Builtins are on that boundary between things we officially support and things we don't officially support. IMO, if we publicly document these particular builtins, we're telling users they're something we support direct calls to and we should aim for higher QoI. But we don't document these (at least that I've found), so the sharp edges are a bit more reasonable. That said, I think we should get the known crashes replaced with diagnostics so long as it isn't significant effort to do so.

**Comment by Endilll** [comment]
@AaronBallman what do you think of the diagnostic I proposed? I wonder if this approach has been tried before and proven to be too fragile to deploy.

**Comment by Endilll** [comment]
Are we going to rely on `-isystem` to suppress false-positives in standard library headers?
